### PR TITLE
[VL] Avoid using WriteFilesSpec which is not serialzable

### PR DESCRIPTION
**Pull request by jackylee-ch**
## What changes were proposed in this pull request?
`WriteFilesSpec` is not serializable, we should not use it in RDDS

## How was this patch tested?
GA

### CODE REVIEW

File: backends-velox/src/main/scala/org/apache/spark/sql/execution/VeloxColumnarWriteFilesExec.scala [hunk 1]
```diff
@@ -90,7 +90,8 @@ case class VeloxWriteFilesMetrics(
  */
 class VeloxColumnarWriteFilesRDD(
     var prev: RDD[ColumnarBatch],
-    writeFilesSpec: WriteFilesSpec,
```
Comment (zhztheplayer) [inline backends-velox/src/main/scala/org/apache/spark/sql/execution/VeloxColumnarWriteFilesExec.scala:93]: Just out of my curiosity, is `writeFilesSpec` actually being serialized? Isn't it just a constructor argument?

**Review by zhztheplayer — APPROVED** [review]

**Comment by ulysses-you** [comment]
Why `WriteFilesSpec` is not serializable, does not it extend `case class` ?

**Comment by jackylee-ch** [comment]
> Why `WriteFilesSpec` is not serializable, does not it extend `case class` ?

@ulysses-you  The `concurrentOutputWriterSpecFunc` in `WriteFilesSpec` is not serializable.

**Comment by ulysses-you** [comment]
Are you sure `concurrentOutputWriterSpecFunc` is not serializable?
image

**Comment by jackylee-ch** [comment]
> Are you sure `concurrentOutputWriterSpecFunc` is not serializable?

`concurrentOutputWriterSpecFunc` is a function, it would seek upper class to find Serializable. Bellow is the problem we met before this pr fixed.
```
	- object not serializable (class: org.apache.spark.SparkContext, value: org.apache.spark.SparkContext@31998d54)
	- field (class: org.apache.spark.sql.hivexxt)
	- object (class org.apache.spark.sql.hive.xx)
	- element of array (index: 2)
	- array (class [Ljava.lang.Object;, size 4)
	- field (class: java.lang.invoke.SerializedLambda, name: capturedArgs, type: class [Ljava.lang.Object;)
	- object (class java.lang.invoke.SerializedLambda, SerializedLambda[capturingClass=class orxx)
	- object (class org.apache.spark.sql.execution.datasources.UserFileFormatWriter$$$Lambda$2964/719020114, org.apache.spark.sql.execution.datasources.UserFileFormatWriter$$$Lambda$2964/719020114@31839df7)
	- field (class: org.apache.spark.sql.execution.datasources.WriteFilesSpec, name: concurrentOutputWriterSpecFunc, type: interface scala.Function1)
	- object (class org.apache.spark.sql.execution.datasources.WriteFilesSpec, WriteFilesSpec(org.apache.spark.sql.execution.datasources.WriteJobDescription@30c77b5a,org.apache.spark.sql.execution.datasources.SQLHadoopMapReduceCommitProtocol@3f3c5251,org.apache.spark.sql.execution.datasources.UserFileFormatWriter$$$Lambda$2964/719020114@31839df7))
	- field (class: org.apache.spark.sql.execution.VeloxColumnarWriteFilesRDD, name: writeFilesSpec, type: class org.apache.spark.sql.execution.datasources.WriteFilesSpec)
	- object (class org.apache.spark.sql.execution.VeloxColumnarWriteFilesRDD, VeloxColumnarWriteFilesRDD[18] at saveAsTable at SparkJobRunner.scala:342)
	- field (class: scala.Tuple2, name: _1, type: class java.lang.Object)
	- object (class scala.Tuple2, (VeloxColumnarWriteFilesRDD[18] at saveAsTable at SparkJobRunner.scala:342,org.apache.spark.sql.execution.datasources.UserFileFormatWriter$$$Lambda$3087/1542323941@7df0da43))

**Comment by ulysses-you** [comment]
@jackylee-ch I think it's caused by your internal changes... Vanilla Spark does not hold spark context in concurrentOutputWriterSpecFunc. You should add `@transient` before sparkcontext field.

**Comment by jackylee-ch** [comment]
> @jackylee-ch I think it's caused by your internal changes... Vanilla Spark does not hold spark context in concurrentOutputWriterSpecFunc. You should add `@transient` before sparkcontext field.

Yes, agree with you. However I think we should avoid use the `concurrentOutputWriterSpecFunc` since we won't use it in RDD.

**Comment by ulysses-you** [comment]
I do not against this change as it is a code improvement. Just make it clear that this pr title and description should not related to serialization...

**Comment by jackylee-ch** [comment]
> I do not against this change as it is a code improvement. Just make it clear that this pr title and description should not related to serialization...

Okey, got it. I have updated the description for more details.

**Pull request merged**